src/Symfony/Component/DependencyInjection/Definition.php

```diff
@@ -830,6 +832,56 @@ public function isAbstract()
     }
 
     /**
+     * Whether this definition is deprecated, that means it should not be called
+     * anymore.
+     *
+     * @param bool   $boolean
+     * @param string $template Template message to use if the definition is deprecated
+     *
+     * @return Definition the current instance
+     *
+     * @todo Support a suggested definition instead ?
+     *
+     * @api
+     */
+    public function setDeprecated($boolean = true, $template = 'The service %s relies on a deprecated definition. You should stop using it.')
+    {
+        $this->deprecated = array(
+            'status' => (bool) $boolean,
+            'template' => $template
+        );
+
+        return $this;
+    }
+
+    /**
+     * Whether this definition is deprecated, that means it should not be called
+     * anymore.
+     *
+     * @return bool
+     *
+     * @api
+     */
+    public function isDeprecated()
+    {
+        return isset($this->deprecated['status']) ? (bool) $this->deprecated['status'] : false;
+    }
+
+    /**
+     * Message to use if this definition is deprecated.
+     *
+     * @param string $id Service id relying on this definition
+     *
+     * @return string
+     *
+     * @api
+     */
+    public function getDeprecationMessage($id)
+    {
+        return sprintf($this->deprecated['template'], $id);
```

Comment:
AFAIK you should check if the key `template` exists, if someone calls this method in a non deprecated definition the key doesn't exists.

```
return strtr($this->deprecated['template'])
```